driver-core/src/main/java/com/datastax/driver/core/Statement.java

```diff
@@ -326,4 +326,49 @@ public Statement setPagingState(PagingState pagingState) {
     ByteBuffer getPagingState() {
         return pagingState;
     }
+
+    /**
+     * Sets whether this statement is idempotent.
+     * <p>
+     * See {@link #isIdempotent()} for more explanations about this property.
+     *
+     * @param idempotent the new value.
+     * @return this {@code Statement} object.
+     */
+    public Statement setIdempotent(boolean idempotent) {
+        this.idempotent = idempotent;
+        return this;
+    }
+
+    /**
+     * Whether this statement is idempotent, i.e. whether it can be applied multiple times
+     * without changing the result beyond the initial application.
+     * <p>
+     * Idempotency plays a role in {@link com.datastax.driver.core.policies.SpeculativeExecutionPolicy speculative executions}.
+     * If a statement is <em>not idempotent</em>, the driver will not schedule speculative
+     * executions for it.
+     * <p>
+     * Note that this method can return {@code null}, in which case the driver will default to
+     * {@link QueryOptions#getDefaultIdempotency()}.
+     * <p>
+     * By default, this method returns {@code null} for all statements, except for
+     * {@link BuiltStatement}s, where the value will be inferred from the query: if it updates
+     * counters or prepends/appends to a list, the result will be {@code false}, otherwise it
+     * will be {@code true}. In all cases, calling {@link #setIdempotent(boolean)} forces a
+     * value that overrides every other mechanism.
+     *
+     * @return whether this statement is idempotent, or {@code null} to use
+     * {@link QueryOptions#getDefaultIdempotency()}.
+     */
+    public Boolean isIdempotent() {
```

Comment:
> null values sound strange and conceptually would mean "I don't know if I'm idempotent"

That's pretty much what it is. The driver does not get in the business of parsing queries, so given a `SimpleStatement` it doesn't know if it's idempotent or not.

> Idempotency is more an internal technical aspect in the driver

Actually I disagree. Idempotency is a intrinsic concept and I think it's quite understandable. And we might end up using in other places than speculative execution.